llvm/lib/Analysis/ConstantFolding.cpp

```diff
@@ -2422,6 +2423,23 @@ static Constant *ConstantFoldScalarCall1(StringRef Name,
 
       return ConstantFP::get(Ty->getContext(), Val);
     }
+
+    case Intrinsic::amdgcn_s_bitreplicate: {
+      uint64_t Val = Op->getZExtValue();
+      uint64_t ReplicatedVal = 0;
+      uint64_t ReplicatedOnes = 0b11;
+      // Input operand is always b32
+      for (unsigned i = 0; i < 32; ++i, ReplicatedOnes <<= 2, Val >>= 1) {
+        uint64_t Bit = Val & 1;
+
+        if (!Bit)
+          continue;
+
+        ReplicatedVal |= ReplicatedOnes;
+      }
```

Comment:
My initial version looked more straightforward like this:

```cpp
uint64_t ReplicatedVal = 0;

for (unsigned i = 0; i < 32; ++i) {
  uint64_t Bit = (Val >> i) & 1;

  if (!Bit)
    continue;

  ReplicatedVal |= (1 << (i * 2));
  ReplicatedVal |= (1 << ((i * 2) + 1);
}
```

But I realized we don't need to shift the values by `i` every iteration, since `i` increases by 1 every time. So instead, I shift the original value by 1 every iteration and shift a mask with two ones (`0b11`) by 2.
The current version might be harder to read or understand. Let me know what you think.

Reply:
This is fun - there are lots of different ways to write it! For example:
```
  for (I = 0; I < 32; ++I)
    ReplicatedVal |= ((Val & (1 << I)) * 3) << I;
```
Or:
```
  Val = (Val & 0x000000000000FFFF) | (Val & 0x00000000FFFF0000) << 16;
  Val = (Val & 0x000000FF000000FF) | (Val & 0x0000FF000000FF00) << 8;
  Val = (Val & 0x000F000F000F000F) | (Val & 0x00F000F000F000F0) << 4;
  Val = (Val & 0x0303030303030303) | (Val & 0x0C0C0C0C0C0C0C0C) << 2;
  Val = (Val & 0x5555555555555555) | (Val & 0xAAAAAAAAAAAAAAAA) << 1;
  ReplicatedVal = Val | Val << 1; // or "Val * 3"
```
But I think it is your decision how you want to write it - I don't mind too much.